src/transcript.rs

```diff
@@ -0,0 +1,62 @@
+//! Defines a `TranscriptProtocol` trait for using a Merlin transcript.
+
+use byteorder::{ByteOrder, LittleEndian};
+use curve25519_dalek::ristretto::CompressedRistretto;
+use curve25519_dalek::scalar::Scalar;
+use merlin::Transcript;
+
+pub trait TranscriptProtocol {
+    /// Commit a domain separator for an `n`-bit, `m`-party range proof.
+    fn rangeproof_domain_sep(&mut self, n: u64, m: u64);
+    /// Commit a domain separator for a length-`n` inner product proof.
+    fn innerproduct_domain_sep(&mut self, n: u64);
+    /// Commit a domain separator for a circuit.
+    fn circuit_domain_sep(&mut self, n: u64, m: u64, q: u64);
+    /// Commit a `scalar` with the given `label`.
+    fn commit_scalar(&mut self, label: &'static [u8], scalar: &Scalar);
+    /// Commit a `point` with the given `label`.
+    fn commit_point(&mut self, label: &'static [u8], point: &CompressedRistretto);
+    /// Compute a `label`ed challenge variable.
+    fn challenge_scalar(&mut self, label: &'static [u8]) -> Scalar;
+}
+
+fn le_u64(value: u64) -> [u8; 8] {
+    let mut value_bytes = [0u8; 8];
+    LittleEndian::write_u64(&mut value_bytes, value);
+    value_bytes
+}
+
+impl TranscriptProtocol for Transcript {
+    fn rangeproof_domain_sep(&mut self, n: u64, m: u64) {
+        self.commit_bytes(b"dom-sep", b"rangeproof");
+        self.commit_bytes(b"n", &le_u64(n));
```

Comment:
Would it be cleaner to avoid `"dom-sep"` and instead do something like this?

```
fn rangeproof_domain(n,m) {
    commit("RangeProof.n", n)
    commit("RangeProof.m", m)
}

fn innerproduct_domain(n) {
    commit("InnerProductProof", n)
}
```

Reply:
I think it's cleaner to keep them as different messages, since it means there's one message to ID the protocol, plus as many other messages as needed to fill in the parameters.